### PR TITLE
Support LaTeX and Math formatting

### PR DESCRIPTION
**Pull request by colombod**
add support for latext strings (#585)

image

### CODE REVIEW

File: Microsoft.DotNet.Interactive/FormattedValue.cs [hunk 1]
```diff
@@ -40,15 +40,12 @@ public static IReadOnlyCollection<FormattedValue> FromObject(object value)
 
         private static string MimeTypeFor(Type returnValueType)
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/FormattedValue.cs]: This whole method should really move into `Formatter` and be configurable so that special knowledge of the types is configurable by extension authors.

File: Microsoft.DotNet.Interactive.Formatting/Formatter.cs [hunk 2]
```diff
@@ -173,6 +173,12 @@ public static void ResetToDefault()
 
             _mimeTypesByType.Clear();
             ConfigureDefaultPlainTextFormattersForSpecialTypes();
+
+            Formatter<LaTeX>.Register((laTex, writer) =>
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Formatting/Formatter.cs]: We have default registrations in other places, e.g. `DefaultPlainTextFormatterSet`. Let's try to keep mime type-specific logic out of `Formatter`.

File: Microsoft.DotNet.Interactive.Formatting/LaTeX.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft. All rights reserved. 
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DotNet.Interactive.Formatting
+{
+    public class LaTeX
+    {
+        private readonly string _latexCode;
+
+        public LaTeX(string latexCode)
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Formatting/LaTeX.cs]: Null check.

File: Microsoft.DotNet.Interactive.Tests/LanguageKernelFormattingTests.cs [hunk 4]
```diff
@@ -89,6 +89,34 @@ public LanguageKernelFormattingTests(ITestOutputHelper output) : base(output)
...
+        public async Task LaTeX_is_rendered_as_text_latex(
+            Language language,
+            string submission,
+            string expectedContent)
+        {
+            var kernel = CreateKernel(language);
+
+            var result = await kernel.SendAsync(new SubmitCode(submission));
+
+            var valueProduced = await result
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Tests/LanguageKernelFormattingTests.cs]: This await and timeout pattern should be avoided. It's a bit verbose and it can hang the test run if someone forgets the timeout trivia.

Do this instead:

```csharp
KernelEvents
  .Should()
  .ContainSingle<DisplayEventBase>()
  .Which
  .FormattedValues
  .Should()
                .ContainSingle(v =>
                    v.MimeType == "text/latex" &&
                    v.Value.ToString().Contains(expectedContent));
```

File: Microsoft.DotNet.Interactive/Kernel.cs [hunk 5]
```diff
@@ -21,9 +21,14 @@ public static class Kernel
         {
             var displayId = DisplayIdGenerator?.Invoke() ?? Guid.NewGuid().ToString();
 
-            if (value is string)
+            switch (value)
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/Kernel.cs]: This also should move to someplace in the formatter infrastructure, where people can register override mime types for specific types.

File: Microsoft.DotNet.Interactive/FormattedValue.cs [hunk 6]
```diff
@@ -28,22 +29,30 @@ public static IReadOnlyCollection<FormattedValue> FromObject(object value)
         {
             var type = value?.GetType();
 
-            var mimeType = MimeTypeFor(type);
+            var mimeTypes = MimeTypesFor(type).Distinct();
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/FormattedValue.cs]: It seems like the mime types uniqueness should be enforced at the registration level.

File: Microsoft.DotNet.Interactive/IKernel.cs [hunk 7]
```diff
@@ -15,6 +15,8 @@ public interface IKernel : IDisposable
     {
         string Name { get; }
 
+        IKernel Parent { get; set; }
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/IKernel.cs]: This doesn't belong on the interface. The hierarchy is really an implementation detail of the `CompositeKernel`.

File: Microsoft.DotNet.Interactive.Formatting/Formatter.cs [hunk 8]
```diff
@@ -19,6 +19,12 @@ public static class Formatter
         private static int _recursionLimit;
         internal static readonly RecursionCounter RecursionCounter = new RecursionCounter();
 
+        private static string _defaultMimeType = HtmlFormatter.MimeType;
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Formatting/Formatter.cs]: I'm not sure this is the correct default.

File: Microsoft.DotNet.Interactive.CSharp/CSharpKernelExtensions.cs [hunk 9]
```diff
@@ -216,8 +217,8 @@ public static string GetDisplayValueId(PackageReference package)
                             context.Publish(new PackageAdded(resolvedReference));
 
                             // Load extensions
-                            await context.HandlingKernel.SendAsync(
-                                new LoadExtensionsInDirectory(
+                            await context.HandlingKernel.GetRoot().SendAsync(
+                                new LoadKernelExtensionsInDirectory(
                                     resolvedReference.PackageRoot,
                                     addedAssemblyPaths));
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.CSharp/CSharpKernelExtensions.cs]: We're doing away with the assembly paths, yeah?

File: Microsoft.DotNet.Interactive.Formatting/Formatter.cs [hunk 10]
```diff
@@ -19,6 +19,12 @@ public static class Formatter
...
+
+        private static readonly ConcurrentDictionary<Type, string> _mimeTypesByType = new ConcurrentDictionary<Type, string>();
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Formatting/Formatter.cs]: Is this being used?

File: Microsoft.DotNet.Interactive.Formatting/Formatter.cs [hunk 11]
```diff
@@ -311,6 +356,11 @@ public static void ResetToDefault()
             SingleLinePlainTextFormatter.WriteEndSequence(writer);
         }
 
+        public static IEnumerable<string> MimeTypesFor(Type type)
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Formatting/Formatter.cs]: This is a bit leaky conceptually because formatters can register lazily.

File: Microsoft.DotNet.Interactive.Formatting/Formatter{T}.cs [hunk 12]
```diff
@@ -71,6 +73,7 @@ void Initialize()
         /// </summary>
         /// <param name="obj">The obj.</param>
         /// <param name="writer">The writer.</param>
+        /// <param name="mimeType">The mimeType</param>
         public static void FormatTo(
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Formatting/Formatter{T}.cs]: Let's not add XML documentation until and unless it's meaningful.

File: Microsoft.DotNet.Interactive.Jupyter.KernelExtensions/Formatting/LaTeX.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) Microsoft. All rights reserved. 
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.DotNet.Interactive.Jupyter.Formatting
+{
+    public class LaTeX
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Jupyter.KernelExtensions/Formatting/LaTeX.cs]: Maybe `LaTeXString`?

File: Microsoft.DotNet.Interactive.Jupyter.Tests/JupyterRequestHandlerTestBase{T}.cs [hunk 14]
```diff
@@ -16,8 +17,8 @@ public abstract class JupyterRequestHandlerTestBase<T> : IDisposable
         where T : Message
     {
         private readonly CompositeDisposable _disposables = new CompositeDisposable();
-        private readonly CSharpKernel _cSharpKernel;
-        private readonly FSharpKernel _fSharpKernel;
+        private readonly IKernel _cSharpKernel;
+        private readonly IKernel _fSharpKernel;
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Jupyter.Tests/JupyterRequestHandlerTestBase{T}.cs]: Why this change?

File: Microsoft.DotNet.Interactive/Extensions/KernelExtensions.cs [hunk 15]
```diff
@@ -23,38 +37,39 @@ public static class KernelExtensions
...
-                    }
+                if (recursive)
+                {
+                    subKernel.VisitSubKernels(onVisit, recursive: true);
                 }
             }
         }
 
-        public static async Task VisitSubkernelsAsync(
+        public static async Task VisitSubKernelsAsync(
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/Extensions/KernelExtensions.cs]: The casing change is incorrect. The `K` should be lower-case.

File: Microsoft.DotNet.Interactive/KernelHierarchy.cs [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.DotNet.Interactive
+{
+    internal static class KernelHierarchy
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/KernelHierarchy.cs]: Are you sure we need this? It's typically going to be more robust to traverse the tree on demand (which sending a command into the root kernel already does) rather than maintain a separate representation of it.

File: dotnet-interactive/dotnet-interactive.csproj [hunk 17]
```diff
@@ -16,6 +16,7 @@
     <Description>.NET Interactive</Description>
     <NoWarn>$(NoWarn);8002</NoWarn><!-- Markdig isn't strongly signed -->
     <NoWarn>$(NoWarn);NU5129</NoWarn><!-- Improper warning about missing props file.  See https://github.com/NuGet/Home/issues/8627 -->
+    <UserSecretsId>all</UserSecretsId>
```
Comment (jonsequitur) [inline dotnet-interactive/dotnet-interactive.csproj]: What's this for?

File: dotnet-interactive/JupyterFormattingKernelExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+using Microsoft.DotNet.Interactive.Commands;
+using Microsoft.DotNet.Interactive.CSharp;
+using Microsoft.DotNet.Interactive.FSharp;
+using Microsoft.DotNet.Interactive.Jupyter.Formatting;
+
+namespace Microsoft.DotNet.Interactive.App
+{
+    public static class KernelFormattingExtensions
+    {
+        public static IKernel UseJupyterFormatting(this CSharpKernel kernel)
```
Comment (jonsequitur) [inline dotnet-interactive/JupyterFormattingKernelExtensions.cs]: The return type should be `CSharpKernel`.

File: dotnet-interactive/JupyterFormattingKernelExtensions.cs [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+using {typeof(LaTeX).Namespace};
+"))).Wait();
+
+            return kernel;
+        }
+
+        public static IKernel UseJupyterFormatting(this FSharpKernel kernel)
+        {
+            if (kernel == null)
+            {
```
Comment (jonsequitur) [inline dotnet-interactive/JupyterFormattingKernelExtensions.cs]: The return type should be `FSharpKernel`.

File: Microsoft.DotNet.Interactive.Formatting/Formatter.cs (hunk 8 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive.Formatting/Formatter.cs]: what should we set then?

File: dotnet-interactive/JupyterFormattingKernelExtensions.cs [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+using System;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Interactive.Commands;
+using Microsoft.DotNet.Interactive.CSharp;
+using Microsoft.DotNet.Interactive.FSharp;
+using Microsoft.DotNet.Interactive.Jupyter.Formatting;
+
+namespace Microsoft.DotNet.Interactive.App
+{
+    public static class KernelFormattingExtensions
```
Comment (jonsequitur) [inline dotnet-interactive/JupyterFormattingKernelExtensions.cs]: I think these should be in `Microsoft.DotNet.Interactive.Jupyter`.

File: Microsoft.DotNet.Interactive.Formatting/Formatter.cs (hunk 10 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive.Formatting/Formatter.cs]: rebase error

File: dotnet-interactive/JupyterFormattingKernelExtensions.cs (hunk 20 above)
Comment (jonsequitur) [reply dotnet-interactive/JupyterFormattingKernelExtensions.cs]: ...and based on the implementation, `UseLaTeX` would be more accurate and more granular.

File: Microsoft.DotNet.Interactive/Extensions/KernelExtensions.cs (hunk 15 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive/Extensions/KernelExtensions.cs]: fixing it

File: dotnet-interactive/dotnet-interactive.csproj (hunk 17 above)
Comment (colombod) [reply dotnet-interactive/dotnet-interactive.csproj]: no idea

File: dotnet-interactive/JupyterFormattingKernelExtensions.cs (hunk 18 above)
Comment (colombod) [reply dotnet-interactive/JupyterFormattingKernelExtensions.cs]: done

File: dotnet-interactive/JupyterFormattingKernelExtensions.cs (hunk 19 above)
Comment (colombod) [reply dotnet-interactive/JupyterFormattingKernelExtensions.cs]: done

File: dotnet-interactive/JupyterFormattingKernelExtensions.cs (hunk 20 above)
Comment (colombod) [reply dotnet-interactive/JupyterFormattingKernelExtensions.cs]: done

File: Microsoft.DotNet.Interactive.Jupyter.Tests/JupyterRequestHandlerTestBase{T}.cs (hunk 14 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive.Jupyter.Tests/JupyterRequestHandlerTestBase{T}.cs]: reverted

File: Microsoft.DotNet.Interactive.Jupyter.KernelExtensions/Formatting/LaTeX.cs (hunk 13 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive.Jupyter.KernelExtensions/Formatting/LaTeX.cs]: sounds right to me

File: Microsoft.DotNet.Interactive/KernelHierarchy.cs (hunk 16 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive/KernelHierarchy.cs]: That class is used to track the root kernel

File: Microsoft.DotNet.Interactive.CSharp/CSharpKernelExtensions.cs (hunk 9 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive.CSharp/CSharpKernelExtensions.cs]: we will only require the packageRoot once we finish

File: Microsoft.DotNet.Interactive/FormattedValue.cs (hunk 6 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive/FormattedValue.cs]: gone

File: Microsoft.DotNet.Interactive.Formatting/Formatter.cs (hunk 11 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive.Formatting/Formatter.cs]: mmm

File: Microsoft.DotNet.Interactive.Formatting/Formatter{T}.cs (hunk 12 above)
Comment (colombod) [reply Microsoft.DotNet.Interactive.Formatting/Formatter{T}.cs]: done
